### PR TITLE
Filter node groups only once, use for loop on filtered node groups

### LINKED ISSUE #1197: Filter out node groups only once
Currently, `NodeGroupFilter.ForEach` is used to perform any operations on matching node groups after passing them through include and exclude globs. This incurs the cost of the match operations multiple times and requires the rest of the code to be aware of the presence of a node group filter. Moreover, this requires an additional error check at the end of the `ForEach` call as any errors from the `ForEach` callback need to be propagated to the caller.

Instead, if the node groups are filtered only once and a matching list of node groups is returned, the match operations will be performed only once and the rest of the code can operate on the resulting slice of node groups directly.

### PR DESCRIPTION
**Pull request by cPu1**
### Description
Closes #1197 

### Checklist

- [x] Code compiles correctly (i.e `make build`)
- [ ] Added tests that cover your change (if possible)
- [x] All unit tests passing (i.e. `make test`)
- [x] Manually tested

### CODE REVIEW

File: pkg/ctl/cmdutils/nodegroup_filter.go [hunk 1]
```diff
@@ -100,6 +100,17 @@ func (f *NodeGroupFilter) MatchAll(nodeGroups []*api.NodeGroup) (sets.String, se
 	return f.doMatchAll(f.collectNames(nodeGroups))
 }
 
+// FilterMatching matches names against the filter and returns all included node groups
+func (f *NodeGroupFilter) FilterMatching(nodeGroups []*api.NodeGroup) []*api.NodeGroup {
+	var match []*api.NodeGroup
+	for _, ng := range nodeGroups {
```
[the remark this answers is not in the document]
Comment (cPu1) [reply pkg/ctl/cmdutils/nodegroup_filter.go]: The issue with that though is `MatchAll` first collects all node group names in a slice first and then constructs two sets with both matching and non-matching items. We only need to process the matching items here and we do that in a single for loop.

Also, I think you meant:
```golang
included, _ := f.MatchAll()
for _, ng := included.List()
```

File: pkg/cfn/manager/delete_tasks.go [hunk 2]
```diff
@@ -45,20 +44,27 @@ func (c *StackCollection) NewTasksToDeleteClusterWithNodeGroups(wait bool, clean
...
-func (c *StackCollection) NewTasksToDeleteNodeGroups(onlySubset sets.String, wait bool, cleanup func(chan error, string) error) (*TaskTree, error) {
+// NewTasksToDeleteNodeGroups defines tasks required to delete all of the nodegroups
+func (c *StackCollection) NewTasksToDeleteNodeGroups(nodeGroups []*api.NodeGroup, wait bool, cleanup func(chan error, string) error) (*TaskTree, error) {
 	nodeGroupStacks, err := c.DescribeNodeGroupStacks()
 	if err != nil {
 		return nil, err
 	}
 
 	tasks := &TaskTree{Parallel: true}
+	hasNodeGroup := func(name string) bool {
```
[the remark this answers is not in the document]
Comment (cPu1) [reply pkg/cfn/manager/delete_tasks.go]: Do we have other places in code where we'd need this?

File: pkg/cfn/manager/delete_tasks.go [hunk 3]
```diff
@@ -4,8 +4,7 @@ import (
 	"fmt"
 
 	"github.com/aws/aws-sdk-go/service/cloudformation"
-
-	"k8s.io/apimachinery/pkg/util/sets"
+	api "github.com/weaveworks/eksctl/pkg/apis/eksctl.io/v1alpha5"
```
[the remark this answers is not in the document]
Comment (cPu1) [reply pkg/cfn/manager/delete_tasks.go]: I can do that, but I should note that it's not the default behaviour of goimports or gofmt.

[the remark this answers is not in the document]
Comment (errordeveloper) [reply pkg/cfn/manager/delete_tasks.go]: I think there is a way to configure it...

File: pkg/ctl/cmdutils/nodegroup_filter.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (errordeveloper) [reply pkg/ctl/cmdutils/nodegroup_filter.go]: Yeah, sounds like there is no benefit in that case...

File: pkg/cfn/manager/delete_tasks.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (errordeveloper) [reply pkg/cfn/manager/delete_tasks.go]: I cannot think of anything, perhaps it'd be premature :)

File: pkg/cfn/manager/delete_tasks.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (errordeveloper) [reply pkg/cfn/manager/delete_tasks.go]: And for me vscode doesn't undo spaces...

[the remark this answers is not in the document]
Comment (cPu1) [reply pkg/cfn/manager/delete_tasks.go]: Yeah, goimports and gofmt maintain any extra newlines added manually but when automatically adding missing imports, goimports will not add a newline for imports in the same group.

**Pull request merged**